### PR TITLE
Create a bundle step on prepublish.

### PR DESCRIPTION
**Pull request by JAForbes**
#442

### CODE REVIEW

**Comment by davidchambers** [comment]
Thanks for getting the ball rolling, @JAForbes! I'll leave some comments. :)

**Comment by JAForbes** [comment]
Thanks @davidchambers 

Few early notes

- The Readme content is just scaffolding, I'd love some thoughts
- Not familiar with XYZ (or even Makefiles :D ) so let me know if things are in the wrong place
- I noticed browserify was already installed, took the liberty of including uglify for minification
- The unpkg links won't work until we publish a release, which is a little bit annoying.  Maybe we should update the readme after first merging the bundle publish step?
- I'm on a windows machine today so I could test this via make, or xyz, I just ran the commands in git bash to test.

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 /coverage/
 /node_modules/
+/dist
```
Comment (davidchambers) [inline .gitignore]: Please add a trailing slash, and preserve alphabetical order.

File: .npmignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+/coverage/
+/node_modules/
+/test/
```
Comment (davidchambers) [inline .npmignore]: Is this file necessary? We specify `"files"` in __package.json__.

Comment (JAForbes) [reply .npmignore]: In that case no 😆

File: Makefile [hunk 3]
```diff
@@ -62,6 +62,11 @@ lint:
 release-major release-minor release-patch:
 	@$(XYZ) --increment $(@:release-%=%)
 
+.PHONY: bundle
+bundle:
+	mkdir dist || echo "Directory already exists"
+	browserify index.js > dist/sanctuary.js
```
Comment (davidchambers) [inline Makefile]: Is Browserify necessary? If so, why is it necessary?

File: README.md [hunk 4]
```diff
@@ -8,6 +8,29 @@ Sanctuary is a JavaScript functional programming library inspired by
...
+
+You can also try sanctuary in your browser right now in the [Ramda REPL](https://goo.gl/pUU9Ye)
+
+If you ever need any help getting started, the sanctuary community is always happy to help.
+
+We're an inclusive community, every question is valid and every level of experience is welcome.  [Chat with us!](https://gitter.im/sanctuary-js/sanctuary)
+
+
+## Sanctuary
+
```
Comment (davidchambers) [inline README.md]: Please move these additions to __index.js__.

File: Makefile (hunk 3 above)
Comment (JAForbes) [reply Makefile]: It means you include one script tag, and sanctuary-def, etc are all bundled together.  I think the main reason you'd want this is for demoing/testing out the library with minimal friction.

File: package.json [hunk 5]
```diff
@@ -33,12 +34,15 @@
     "remember-bower": "0.1.x",
     "sanctuary-style": "0.5.x",
     "transcribe": "1.0.x",
+    "uglify-js": "^3.1.0",
```
Comment (davidchambers) [inline package.json]: Please use `"3.1.x"` here.

File: Makefile (hunk 3 above)
Comment (davidchambers) [reply Makefile]: I'm certainly in favour of providing a convenient `sanctuary+dependencies` bundle. It seems we're using Browserify here for dependency resolution. Is this correct? If so, I'm hoping something simpler would suffice:

```
cat \
  node_modules/sanctuary-type-identifiers/index.js \
  node_modules/sanctuary-type-classes/index.js \
  node_modules/sanctuary-def/index.js \
  index.js \
  >dist/sanctuary.js
```

File: README.md [hunk 6]
```diff
@@ -8,6 +8,8 @@ Sanctuary is a JavaScript functional programming library inspired by
 [Haskell][] and [PureScript][]. It's stricter than [Ramda][], and
 provides a similar suite of functions.
 
+## Sanctuary
+
```
Comment (davidchambers) [inline README.md]: Remove these lines too. ;)

**Comment by JAForbes** [comment]
FTR Just tested `make bundle` and made an index.html that included `./dist/sanctuary.min.js` and it worked as expected.

File: Makefile (hunk 3 above)
Comment (JAForbes) [reply Makefile]: Yeah that looks good.

File: Makefile [hunk 7]
```diff
@@ -62,6 +64,11 @@ lint:
 release-major release-minor release-patch:
 	@$(XYZ) --increment $(@:release-%=%)
 
+.PHONY: bundle
+bundle:
+	mkdir dist || echo "Directory already exists"
+	$(BROWSERIFY) --standalone S index.js > dist/sanctuary.js
+	cat dist/sanctuary.js | $(UGLIFYJS) -mc > dist/sanctuary.min.js
```
Comment (davidchambers) [inline Makefile]: Here's the idiomatic way to express the dependencies:

```make
.PHONY: bundle
bundle: dist/sanctuary.js dist/sanctuary.min.js

.PHONY: dist/sanctuary.js
dist/sanctuary.js:
	mkdir -p -- '$(@D)'
	$(BROWSERIFY) --standalone S index.js >'$@'

dist/sanctuary.min.js: dist/sanctuary.js
	mkdir -p -- '$(@D)'
	cat '$<' | $(UGLIFYJS) -mc >'$@'
```

We could hard-code the dependencies of `dist/sanctuary.js` to avoid making it a phony target.

File: Makefile [hunk 8]
```diff
@@ -3,6 +3,8 @@ ESLINT = node_modules/.bin/eslint --config node_modules/sanctuary-style/eslint-e
 ISTANBUL = node_modules/.bin/istanbul
 NPM = npm
 REMARK = node_modules/.bin/remark --frail --no-stdout
+BROWSERIFY = node_modules/.bin/browserify
+UGLIFYJS = node_modules/.bin/uglifyjs
```
Comment (davidchambers) [inline Makefile]: Please preserve alphabetical order.

File: index.js [hunk 9]
```diff
@@ -43,6 +43,26 @@
 //.
 //. Sanctuary is designed to work in Node.js and in ES5-compatible browsers.
 //.
+//. ## Installation
+//.
+//. Sanctuary can be installed via npm
```
Comment (davidchambers) [inline index.js]: Let's add a colon (`:`) at the end of this sentence.

File: index.js [hunk 10]
```diff
@@ -43,6 +43,26 @@
...
+//.
+//. ```
+//. npm install sanctuary
+//. ```
```
Comment (davidchambers) [inline index.js]: Let's add `console` and a minimal prompt:

    ```console
    $ npm install sanctuary
    ```

File: index.js [hunk 11]
```diff
@@ -43,6 +43,26 @@
...
+//.
+//. Or simply include a script tag in your html page
```
Comment (davidchambers) [inline index.js]: I'd like to see _HTML document_ in place of _html page_. :)

File: index.js [hunk 12]
```diff
@@ -43,6 +43,26 @@
...
+//.
+//. ```html
+//. <script src="https://unpkg.com/sanctuary@latest/dist/sanctuary.min.js"><script>
```
Comment (davidchambers) [inline index.js]: The closing tag is missing its slash (`/`).

File: Makefile [hunk 13]
```diff
@@ -62,6 +63,33 @@ lint:
...
+
+.PHONY: dist/sanctuary.js
+dist/sanctuary.js:
+	mkdir -p -- '$(@D)'
+	cat \
+  node_modules/sanctuary-type-identifiers/index.js \
+  node_modules/sanctuary-type-classes/index.js \
+  node_modules/sanctuary-def/index.js \
+  index.js \
+  >dist/sanctuary.js
```
Comment (davidchambers) [inline Makefile]: If we don't need Browserify we should be able to avoid making `dist/sanctuary.js` a phony target:

```make
dist/sanctuary.js: \
		node_modules/sanctuary-type-identifiers/index.js \
		node_modules/sanctuary-type-classes/index.js \
		node_modules/sanctuary-def/index.js \
		index.js
	mkdir -p -- '$(@D)'
	cat -- $^ >'$@'
```

File: Makefile [hunk 14]
```diff
@@ -62,6 +63,33 @@ lint:
...
+.PHONY: bundle
+bundle:
+	mkdir dist || echo "Directory already exists"
+	cat \
+  node_modules/sanctuary-type-identifiers/index.js \
+  node_modules/sanctuary-type-classes/index.js \
+  node_modules/sanctuary-def/index.js \
+  index.js \
+  >dist/sanctuary.js
+	cat dist/sanctuary.js | $(UGLIFYJS) -mc > dist/sanctuary.min.js
```
Comment (davidchambers) [inline Makefile]: This paragraph can now be deleted since (below) we state that `bundle` depends on `dist/sanctuary.js` and `dist/sanctuary.min.js`.

File: index.js [hunk 15]
```diff
@@ -43,6 +43,26 @@
...
+//. $ npm install sanctuary
+//. ```
+//.
+//. Or simply include a script tag in your html document:
+//.
+//. ```html
+//. <script src="https://unpkg.com/sanctuary@latest/dist/sanctuary.min.js"></script>
+//. ```
+//.
+//. You can also try sanctuary in your browser right now in the [Ramda REPL](https://goo.gl/pUU9Ye)
```
Comment (davidchambers) [inline index.js]: I'm wary of linking to the Ramda REPL because I don't know how frequently its dependencies are updated, and the relationship between Ramda and Sanctuary has been a source of confusion in the past.

File: index.js [hunk 16]
```diff
@@ -43,6 +43,26 @@
...
+//.
+//. If you ever need any help getting started, the sanctuary community is always happy to help.
```
Comment (davidchambers) [inline index.js]: `s/sanctuary/Sanctuary/`

File: index.js [hunk 17]
```diff
@@ -43,6 +43,26 @@
...
+//. Sanctuary can be installed via npm:
+//.
+//. ```console
+//. $ npm install sanctuary
+//. ```
+//.
+//. Or simply include a script tag in your html document:
+//.
+//. ```html
+//. <script src="https://unpkg.com/sanctuary@latest/dist/sanctuary.min.js"></script>
```
Comment (davidchambers) [inline index.js]: For debugging purposes would it be better to suggest the unminified file here?

File: index.js (hunk 15 above)
Comment (JAForbes) [reply index.js]: That's fair, but it is super helpful for demoing the library.  Any ideas for alternatives that have the same benefits?  Maybe repl.it, or runkit?

File: index.js [hunk 18]
```diff
@@ -43,6 +43,26 @@
...
+//. ```
+//.
+//. You can also try sanctuary in your browser right now in the [Ramda REPL](https://goo.gl/pUU9Ye)
+//.
+//. If you ever need any help getting started, the sanctuary community is always happy to help.
+//.
+//. We're an inclusive community, every question is valid and every level of experience is welcome.  [Chat with us!](https://gitter.im/sanctuary-js/sanctuary)
```
Comment (davidchambers) [inline index.js]: Please wrap lines at 79 characters.

File: index.js (hunk 17 above)
Comment (JAForbes) [reply index.js]: We could, I was thinking the target audience for the bundle is unlikely to be debugging sanctuary source code.  So I was optimizing for load time.  What do you think?

**Comment by davidchambers** [comment]
> The unpkg links won't work until we publish a release, which is a little bit annoying. Maybe we should update the readme after first merging the bundle publish step?

The good news is that the readme won't be updated until we publish the next version. :)

**Comment by JAForbes** [comment]
True, I just meant, it's annoying to merge a change that is effectively untested.  We're saying "use this link" but it's an assumption that it works.

File: index.js (hunk 15 above)
Comment (JAForbes) [reply index.js]: Would this be better https://repl.it/Ky1q ?

Comment (davidchambers) [reply index.js]: I leave this decision to you. If you consider the Ramda REPL the best of the three let's use it but let's first understand the process for updating its dependencies. Linking to a REPL running an old version of Sanctuary might do more harm than good.

File: index.js (hunk 17 above)
Comment (davidchambers) [reply index.js]: > I was optimizing for load time.

That seems reasonable to me. :)

**Comment by davidchambers** [comment]
> I just meant, it's annoying to merge a change that is effectively untested.

Oh, I'm with you now. Let's proceed on the assumption that everything will work, but be ready to make a quick patch release.

File: index.js [hunk 19]
```diff
@@ -43,6 +43,30 @@
...
+//. ```
+//.
+//. Or simply include a script tag in your html document:
+//.
+//. ```html
+//. <script src="https://unpkg.com/sanctuary@latest/dist/sanctuary.min.js"></script>
+//. ```
+//.
+//. You can also try Sanctuary in your browser right now in the
+//. [Ramda REPL](https://goo.gl/pUU9Ye)
```
Comment (davidchambers) [inline index.js]: `s/$/./`

File: index.js [hunk 20]
```diff
@@ -43,6 +43,30 @@
...
+//.
+//. If you ever need any help getting started, the sanctuary community is
```
Comment (davidchambers) [inline index.js]: `s/sanctuary/Sanctuary/`

File: index.js [hunk 21]
```diff
@@ -43,6 +43,30 @@
...
+//. always happy to help.
+//.
+//. We're an inclusive community, every question is valid and every level of
```
Comment (davidchambers) [inline index.js]: `s/,/:/`

File: index.js (hunk 15 above)
Comment (davidchambers) [reply index.js]: > Would this be better <https://repl.it/Ky1q> ?

This looks good to me. Does `require('sanctuary')` implicitly refer to the most recently published version?

File: index.js (hunk 19 above)
Comment (JAForbes) [reply index.js]: Sorry not following this one.

Comment (davidchambers) [reply index.js]: This one was cryptic, I'm sorry. The `$` matches end-of-line:

```console
$ echo foo | sed 's/$/./'
foo.
```

File: Makefile (hunk 13 above)
Comment (JAForbes) [reply Makefile]: Is this particular change resolved?

File: Makefile (hunk 14 above)
Comment (JAForbes) [reply Makefile]: Is this particular change resolved?

File: Makefile (hunk 13 above)
Comment (davidchambers) [reply Makefile]: I will push a commit to your branch. ;)

File: index.js [hunk 22]
```diff
@@ -43,6 +43,30 @@
...
+//. ```
+//.
+//. Or simply include a script tag in your html document:
+//.
+//. ```html
+//. <script src="https://unpkg.com/sanctuary@latest/dist/sanctuary.min.js"></script>
+//. ```
+//.
+//. You can also try Sanctuary in your browser right now in the
+//. [Ramda REPL](https://goo.gl/pUU9Ye).
```
Comment (davidchambers) [inline index.js]: Please (temporarily) add this snippet to the pull request to make it easier to review. :)

Comment (davidchambers) [reply index.js]: :zap:
